internal/core/metadata/operators/device/update.go

```diff
@@ -0,0 +1,172 @@
+/*******************************************************************************
+ * Copyright 2019 Dell Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except
+ * in compliance with the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ *******************************************************************************/
+
+package device
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/edgexfoundry/edgex-go/internal/core/metadata/errors"
+	"github.com/edgexfoundry/edgex-go/internal/pkg/db"
+	"github.com/edgexfoundry/go-mod-core-contracts/clients/logger"
+	contract "github.com/edgexfoundry/go-mod-core-contracts/models"
+)
+
+type DeviceUpdateExecutor interface {
+	Execute() (err error)
+}
+
+type updateDevice struct {
+	database DeviceUpdater
+	device   contract.Device
+	events   chan DeviceEvent
+	logger   logger.LoggingClient
+}
+
+func NewUpdateDevice(ch chan DeviceEvent, db DeviceUpdater, d contract.Device, logger logger.LoggingClient) DeviceUpdateExecutor {
+	return updateDevice{database: db, events: ch, device: d, logger: logger}
+}
+
+func (op updateDevice) Execute() (err error) {
+	var evt DeviceEvent
+
+	// Check if the device exists
+	// First try ID
+	var oldDevice contract.Device
+	oldDevice, err = op.database.GetDeviceById(op.device.Id)
+	if err != nil {
+		// Then try name
+		oldDevice, err = op.database.GetDeviceByName(op.device.Name)
+		if err != nil {
+			if err == db.ErrNotFound {
+				err = errors.NewErrItemNotFound(fmt.Sprintf("device not found: %s %s", op.device.Name, op.device.Id))
+			}
+			op.logger.Error(err.Error())
+			evt.Error = err
+			op.events <- evt
+			return
+		}
+	}
+
+	if err = updateDeviceFields(op.device, &oldDevice, op.database); err != nil {
+		op.logger.Error(err.Error())
+		evt.Error = err
+		op.events <- evt
+		return
+	}
+
+	if err = op.database.UpdateDevice(oldDevice); err != nil {
+		op.logger.Error(err.Error())
+		evt.Error = err
+		op.events <- evt
+		return
+	}
+
+	// Device updated successfully.  Publish event onto supplied channel.
+
+	evt.DeviceId = op.device.Id
+	evt.DeviceName = op.device.Name
+	evt.HttpMethod = http.MethodPost
+	evt.ServiceId = op.device.Service.Id
+
+	op.events <- evt
+
+	return nil
+}
+
+func updateDeviceFields(from contract.Device, to *contract.Device, dbClient DeviceUpdater) error {
+
```

Comment:
As I said to @AnthonyMBonafide in my comments on his [PR](https://github.com/edgexfoundry/edgex-go/pull/1498#discussion_r300075041), I know you didn't write this function and it was just brought over from the existing router logic. Please refactor so that the `*to` parameter isn't provided by the caller. Instead, the function should return the new Device instance.